### ISSUE TITLE
Configuration and build setting changes

### ISSUE THREAD

**Issue by yonaskolb** [workshop/struct#132]
This is a proposal of a different way to specify configurations and build settings. It has a few different ideas in it that work together. You can take it or leave it, but let me know what you think. My examples are using the yaml format.

### Separation of configurations and build settings
Overrides are removed from `configurations` and `target.configurations`. A new property called `buildSettings` is created for the project at the top level and a target. 

In each `buildSettings` a list of build settings is defined. For each build setting a simple value can be supplied, or a map of configuration to build setting, allowing for more granularity. When settings are set for specific configurations an optional $base property can be set for a default value, so as not to have to set all of them if just changing one config. The existing behaviour of allowing an array for multiline settings would stay the same, and is allowed at both levels.
```
buildSettings:
    - BUILD_SETTING:
        $base: default value
        debug: debug value
        release:
        - $(inherrited)
        - value
    - BUILD_SETTING2: build value
```

The project configurations is then limited to specifying a name and type.
```
configurations:
  test: debug
  release: release
```
The configurations of a target is then limited to specify a map between a configuration and an xcconfig file:
```
MyTarget:
  configurations:
      debug: debug.xcconfig
      release: release.xcconfig
```
This makes it clear what each section does, and aligns is more with how configurations and build settings are defined and configured in Xcode

### Build Setting Groups
There would be a top level `buildSettingGroups`(name?). This would list a bunch of groups of build settings. These could then be referenced from anywhere where build settings are configured.
This is very similar to the old profiles, but user facing. Each group as the same capabilities as the `buildSettings` sections above

```
buildSettingGroups:
  mygroup1:
    - BUILD_SETTING: build value
    - BUILD_SETTING2: build value
```

### Build setting group references
Wherever build settings are defined (buildSettingGroups, buildSettings, target.buildSettings), an optional list of `groups` can be defined. This merges all the references groups in order, and then merges the specific build settings on top. 
Because groups and build settings are defined in the same element the build settings would have to be moved into a `settings` property.

```
buildSettings:
  groups:
    - myGroup
    - otherGroup
  settings:
    - BUILD_SETTING: build value
```
---
With these changes, settings become much more powerful, granular, and composable. They map more to how these things are defined in xcode, making the mental model easier.
For many cases it also makes xcconfig files reduntant and for settings to be able to be specified just in yaml.

**Comment by lyptt** [comment]
So this essentially is proposing the ability to pre-define your overrides in another section, then reference them in project and target configurations? Essentially expanding out the $base concept.

There are ways to do this right now in YAML natively, however it's not the easiest syntax to understand:

```yaml
.debug-overrides: &debug-overrides
  IPHONEOS_DEPLOYMENT_TARGET: '9.1'

targets:
  My app:
    configurations:
      debug:
        overrides:
          <<: *debug-overrides
```

I'll definitely take this under consideration. Thanks for proposing this!

**Comment by yonaskolb** [comment]
Another possible additional change here would be to remove configurations from target entirely and specify each target xcode configuration files in the top level configurations. 
Additionally an optional project xcconfig file could be applied with `config`
```
configurations:
  debug:
    type: debug
    config: project.xcconfig
    targetConfigs:
      MyTarget: myTargerDebug.xcconfig
```

This change isn't required but could be considered. This would match the Xcode UI even more, making the mental model simpler, and keeps configuration settings all in one place.

**Comment by lyptt** [comment]
In this scenario I think that moving target configuration out of the target definitions would make it more difficult to see at a glance how each target is configured. 

I'm also keen to stick to each block having its own concerns. With the exception of variants, each block could be reused in another spec file with a simple copy paste. 

Fundamentally spec files aren't meant to be a direct clone of xcodeproj files. The ethos behind Struct is to make defining Xcode projects as simple as possible, and to make Xcode fit better into the modern developer's workflow and approach.

**Comment by yonaskolb** [comment]
Yep, makes sense, I actually agree. Just thought I'd put it out there as an option

**Comment by lyptt** [comment]
Thinking about this further, I think based on my previous comment this would add ties between target blocks and the overall spec file that would make things more difficult to maintain. There are existing ways to do this without relying on a spec schema change:

**YAML (not JSON)**
```yaml
.debug-overrides: &debug-overrides
  IPHONEOS_DEPLOYMENT_TARGET: '9.1'

targets:
  My app:
    configurations:
      debug:
        <<: *debug-overrides
```

**Specfile**
```ruby
def add_debug_overrides
  override 'IPHONEOS_DEPLOYMENT_TARGET', '9.1'
end

target 'My app' do
  configuration 'debug' do
    add_debug_overrides
  end
end

```

**Comment by yonaskolb** [comment]
Just as a note splitting up configuration xcconfig files and build setting would also fix #118
